File: 转场动画/TransitionAnimation/ImageAnimation/ViewController.m
//
//  ViewController.m
//  ImageAnimation
//
//  Created by 姜泽东 on 2017/9/6.
//  Copyright © 2017年 MaiTian. All rights reserved.
//

#import "ViewController.h"

#import "MTVCAnimationTransition.h"
#import "MTImageVC.h"
#import "MTTestVCBack.h"
#import "AAPLCustomPresentationController.h"

@implementation MyNav

- (BOOL)shouldAutorotate
{
    return YES;
}

- (UIInterfaceOrientationMask)supportedInterfaceOrientations
{
    return UIInterfaceOrientationMaskAll;
}


@end

@interface ViewController ()<UINavigationControllerDelegate>

@property (weak, nonatomic) IBOutlet UIImageView *imgView;


@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    self.navigationController.delegate = self;
    
    self.image = [UIImage imageNamed:@"2.jpeg"];
    //封装
    
}

// 点击
- (nullable id <UIViewControllerAnimatedTransitioning>)navigationController:(UINavigationController *)navigationController
                                            animationControllerForOperation:(UINavigationControllerOperation)operation
                                                         fromViewController:(UIViewController *)fromVC
                                                           toViewController:(UIViewController *)toVC{

    if (operation != UINavigationControllerOperationNone) {
        MTVCAnimationTransition *animation = [MTVCAnimationTransition new];
        animation.isPush = operation == UINavigationControllerOperationPush;
        return animation;
    }
    return nil;
}

//手势(需要交互的时候)
-(id<UIViewControllerInteractiveTransitioning>)navigationController:(UINavigationController *)navigationController interactionControllerForAnimationController:(id<UIViewControllerAnimatedTransitioning>)animationController {

    return nil;
}

- (UIImage *)image {
    
    return _image;
}

- (CGRect)imageFrame {
    
    return self.imgView.frame;
}

- (UIView *)imageTargetView {
    
    return self.imgView;
}

- (IBAction)jump:(id)sender {
    
    MTImageVC *ani = (MTImageVC *)[self.storyboard instantiateViewControllerWithIdentifier:@"ani"];
    ani.image = self.image;
    ani.modalPresentationStyle = UIModalPresentationFullScreen;
    [self presentViewController:ani animated:YES completion:nil];
//    [self.navigationController pushViewController:ani animated:YES];
//    UIViewController *vc = [NSClassFromString(@"AAPLSwipeFirstViewController") new];
//    vc.modalPresentationStyle = UIModalPresentationFullScreen;
//    [self presentViewController:vc animated:YES completion:nil];
    
}
- (IBAction)customJump:(id)sender {
    
    UIViewController *vc = [NSClassFromString(@"AAPLCustomPresentationSecondViewController") new];
    
    AAPLCustomPresentationController *presentationController NS_VALID_UNTIL_END_OF_SCOPE;
   
    presentationController = [[AAPLCustomPresentationController alloc] initWithPresentedViewController:vc presentingViewController:self];
    
    vc.transitioningDelegate = presentationController;
    
    vc.modalPresentationStyle = UIModalPresentationCustom;
//    vc.modalPresentationStyle = UIModalPresentationOverFullScreen;
    [self presentViewController:vc animated:YES completion:nil];
}

- (BOOL)shouldAutorotate
{
    return YES;
}

- (UIInterfaceOrientationMask)supportedInterfaceOrientations
{
    return UIInterfaceOrientationMaskAll;
}

@end

File: MyAlgorithmEnter/LinkerTable/LinkedTable/ViewController.swift
//
//  ViewController.swift
//  LinkedTable
//
//  Created by 姜泽东 on 2017/12/26.
//  Copyright © 2017年 MaiTian. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    let nodeList = ListTable()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        
        nodeList.appendToTail(1)
        nodeList.appendToTail(5)
        nodeList.appendToTail(3)
        nodeList.appendToTail(2)
        nodeList.appendToTail(4)
        nodeList.appendToTail(2)
        
        //sgetLeftList(nodeList.head, 3)
        print(partition(nodeList.head, 3)!)
        print(nodeList.description)
        print(nodeList.head!.value,nodeList.tail!.value)
    }

    // 提取新链表的头节点
    func getLeftList(_ head: ListNode?, _ x: Int) -> ListNode? {
        let dummy = ListNode(0)
        var pre = dummy
        var node = head
        
        while node != nil {
            if node!.value < x {
                pre.next = node
                pre = node!
            }
            node = node!.next
        }
        return dummy.next
    }
    
    func partition(_ head: ListNode?, _ x: Int) -> ListNode? {
        // 引入Dummy节点
        let prevDummy = ListNode(0)
        var prev = prevDummy
        let postDummy = ListNode(0)
        var post = postDummy
        
        var node = head
        
        // 用尾插法处理左边和右边
        while node != nil {
            if node!.value < x {
                prev.next = node
                prev = node!
            } else { 
                post.next = node
                post = node!
            }
            node = node!.next
        }
        
        // 左右拼接
        post.next = nil
        prev.next = postDummy.next
        
        return prevDummy.next
    }
    
}

